kernel/src/hil/date_time.rs

```diff
@@ -0,0 +1,71 @@
+use crate::ErrorCode;
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum DayOfWeek {
+    Sunday,
+    Monday,
+    Tuesday,
+    Wednesday,
+    Thursday,
+    Friday,
+    Saturday,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum Month {
+    January,
+    February,
+    March,
+    April,
+    May,
+    June,
+    July,
+    August,
+    September,
+    October,
+    November,
+    December,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct Date {
+    pub year: u16,
+    pub month: Month,
+    pub day: u8,
+    pub day_of_week: DayOfWeek,
+    pub hour: u8,
+    pub minute: u8,
+    pub seconds: u8,
+}
+
+/// Interface for reading and setting the current time
+pub trait DateTime<'a> {
+    /// Request driver to return date and time
+    ///
+    /// When successful this function call must be followed by a call
+    /// to `callback_get_date` which provides the actual date and time
+    /// or an error.
+    fn get_date_time(&self) -> Result<(), ErrorCode>;
+
+    /// Sets the current date and time
+    ///
+    /// When successful this function call must be followed by a call
+    /// to `callback_set_date`.
+    fn set_date_time(&self, date_time: Date) -> Result<(), ErrorCode>;
+
+    /// Sets a client that calls the callback function when date and time is requested
+    fn set_client(&self, client: &'a dyn DateTimeClient);
+}
+
+/// Callback handler for when current date is read or set.
+pub trait DateTimeClient {
+    /// Called when a date time reading has completed.
+    /// Takes `Ok(DateTime)` of current date and passes it when scheduling an upcall.
+    /// If an error is encountered it takes an `Err(ErrorCode)`
```

Comment:
What are the valid errors, and what do they mean?

Reply:
The `ErrorCode` passed as parameter is the same `ErrorCode` that the low level driver might issue if it encounters an inconsistent reading. All ErrorCodes are going to be translated into (usize) statuscodes in the `date_time` capsule. So unless the `into_statuscode` function does not support all error codes, there shouldn't be any invalid errors. 

For synchronous drivers (like the *rp2040/src/rtc.rs*) the errors can be checked before setting the deferred call and therefore in the callback function it is assumed that the result given by the driver is always correct. 
Not sure about asynchronous reading and setting. In this case I assume that the values returned by the driver should be checked right before `callback_get_date` is called. I think the errors that would make sense in this case would be : `FAIL`, `INVAL` or `UNINSTALLED`.

Reply:
Those could make sense: what do they mean? We typically try to document the meaning of different `ErrorCode` values similarly to how POSIX documents errno values.